### PR TITLE
Add and display the sentiment score of comments.

### PR DESCRIPTION
**Pull request by ak47na**
Add score instance variable in Comment class and set its value using `getMessageScore(String message)`.
Remove duplicate message in index.html
Update `getComments()` in script.js to get the sentiment score from the JSON and display it in parentheses after the text of the comment.

TODO[ak47na]: Update the login and logout links using the code in branch AnaKDevelopment3.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -157,4 +157,18 @@ private String getNewComment(HttpServletRequest request) throws UnsupportedEncod
     byte[] commentBytes = newComment.getBytes("UTF-8");
     return newComment;
   }
+
+  // TODO[ak47na]: create Nickname class and remove duplicate getUserNickname
```
Comment (alexcepoi) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: FYI: This change seems unrelated to the PR description of displaying sentiment scores.

Should this have been a separate PR?

It's also unused, so I'm a bit confused as to why it's included here.

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 2]
```diff
@@ -14,14 +14,31 @@
...
     this.message = message;
     this.userData = userData;
+    this.score = getMessageScore(message);
+  }
+
+  private float getMessageScore(String message) throws IOException {
+    Document doc = Document.newBuilder().setContent(message).setType(Document.Type.PLAIN_TEXT).build();
+    LanguageServiceClient languageService = LanguageServiceClient.create();
+    Sentiment sentiment = languageService.analyzeSentiment(doc).getDocumentSentiment();
+    float score = sentiment.getScore();
```
Comment (alexcepoi) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: The dance you do here around this variable seems to imply that `sentiment` becomes invalid after `languageService.close` is called.

But I'm not quite sure why that would be the case. Would something like this not work?

```
client = LanguageServiceClient.create();
sentiment = ...getDocumentSentiment();
client.close();

return sentiment.getScore();
```

File: portfolio/src/main/webapp/index.html [hunk 3]
```diff
@@ -48,9 +46,10 @@ <h1>Comments:</h1>
 
       <input type="submit" />
       </form>
-
-      <a id="loginLink" href = "/_ah/login?continue=%2F" title="Login here">Login</a>
-      <a id="logoutLink" href = "/_ah/logout?continue=%2F" title="Logout here">Logout</a>
+      
+      <a id="changeNicknameLink" href="/nickname" title = "Change nickname">Change your nickname here</a>
```
Comment (alexcepoi) [inline portfolio/src/main/webapp/index.html]: Where in your java server are requests to "/nickname" handled?

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 4]
```diff
@@ -14,14 +14,31 @@
...
+  private final float score;
 
-  public Comment (String message, String userData) {
+  public Comment(String message, String userData) throws IOException {
     this.message = message;
     this.userData = userData;
+    this.score = getMessageScore(message);
+  }
+
+  private float getMessageScore(String message) throws IOException {
```
Comment (alexcepoi) [inline portfolio/src/main/java/com/google/sps/data/Comment.java:38]: Add a short comment on what this function does and in which cases it throws IOException. This may be unclear for a user who doesn't read your implementation (and in general users should not have to read your entire implementation to understand what a function/class does).

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 2 above)
Comment (alexcepoi) [reply portfolio/src/main/java/com/google/sps/data/Comment.java]: https://github.com/GoogleCloudPlatform/java-docs-samples/blob/master/language/analysis/src/main/java/com/google/cloud/language/samples/Analyze.java#L160

suggests using the try-with-resource format here (so that you don't have to worry about properly closing), and handling the possibility of sentiment being null.

File: portfolio/src/main/webapp/index.html (hunk 3 above)
Comment (ak47na) [reply portfolio/src/main/webapp/index.html]: Requests to "/nickname" are handled in NicknameServlet.java

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 2 above)
Comment (ak47na) [reply portfolio/src/main/java/com/google/sps/data/Comment.java]: I changed the code to use try-with-resource, but the version you mentioned (which closes the client before returning the sentiment score) would also work:

Comment (alexcepoi) [reply portfolio/src/main/java/com/google/sps/data/Comment.java]: Yes, all variants technically work.

The try-with-resource approach is arguably more future-proof. One can imagine a future where the try block expands and now:
```
client = LanguageServiceClient.create();
someFunctionWithCanRaiseException();  // exception got raised (and then handled in the caller).
client.close();  // never gets called.. this is bad.
```

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 4 above)
Comment (alexcepoi) [reply portfolio/src/main/java/com/google/sps/data/Comment.java:38]: Consider documenting the conditions in which IOException is raised, as it's not obvious at all.

e.g.
@throws IOException if fails to establish connection to LanguageService.

https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html#throwstag

File: portfolio/src/main/webapp/index.html (hunk 3 above)
Comment (alexcepoi) [reply portfolio/src/main/webapp/index.html]: Ah gotcha, this is following up on a previous PR.

Fine for now, but please separate unrelated changes in their own PRs in the future.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (alexcepoi) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Sorry this should not have been an FYI, I am genuinely confused :)

I see this used in `createCommentEntity`, was this not compiling before and this fixes it? Can you elaborate a bit?

And try to send these fixes as their own PRs in the future, there is no such thing as a "too small PR" :)

**Review by alexcepoi — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (ak47na) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Yes, the `getUserNickname(String userId)` was not related to the PR. When I merged the previous PR, I accidentally deleted the function. 
I added it in this PR because the code was not compiling and I could not test it.

Thank you for the advice, I will make separete PRs for fixes from now on.

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 4 above)
Comment (ak47na) [reply portfolio/src/main/java/com/google/sps/data/Comment.java:38]: Thank you for the resource, I added @throws for the exception.
